arangod/RestServer/Metric/MetricsFeature.cpp

```diff
@@ -0,0 +1,524 @@
+}
+
+Metric* MetricsFeature::get(MetricKey const& key) {
+  std::lock_guard guard{_lock};
+  auto it = _registry.find(key);
+  if (it == _registry.end()) {
+    return nullptr;
+  }
+  return it->second.get();
+}
+
+bool MetricsFeature::remove(Builder const& builder) {
+  MetricKey key{builder.name(), builder.labels()};
+  std::lock_guard guard{_lock};
+  return _registry.erase(key) != 0;
+}
+
+bool MetricsFeature::exportAPI() const { return _export; }
+
+bool MetricsFeature::exportReadWriteMetrics() const {
+  return _exportReadWriteMetrics;
+}
+
+void MetricsFeature::validateOptions(std::shared_ptr<options::ProgramOptions>) {
+  if (_exportReadWriteMetrics) {
+    serverStatistics().setupDocumentMetrics();
+  }
+}
+
+void MetricsFeature::toPrometheus(std::string& result, bool v2) const {
+  // minimize reallocs
+  result.reserve(32768);
+
+  // QueryRegistryFeature
+  auto& q = server().getFeature<QueryRegistryFeature>();
+  q.updateMetrics();
+  {
+    bool changed = false;
+
+    std::lock_guard guard(_lock);
+    if (_globalLabels.find("shortname") == _globalLabels.end()) {
+      std::string shortName = ServerState::instance()->getShortName();
+      // Very early after a server start it is possible that the
+      // short name is not yet known. This check here is to prevent
+      // that the label is permanently empty if metrics are requested
+      // too early.
+      if (!shortName.empty()) {
+        _globalLabels.try_emplace("shortname", shortName);
+        changed = true;
+      }
+    }
+    if (_globalLabels.find("role") == _globalLabels.end() &&
+        ServerState::instance() != nullptr &&
+        ServerState::instance()->getRole() != ServerState::ROLE_UNDEFINED) {
+      _globalLabels.try_emplace("role", ServerState::roleToString(
+                                            ServerState::instance()->getRole()));
+      changed = true;
+    }
+    if (changed) {
+      bool first = true;
+      for (auto const& i : _globalLabels) {
+        if (!first) {
+          _globalLabelsStr += ",";
+        } else {
+          first = false;
+        }
+        _globalLabelsStr += i.first + "=\"" + i.second + "\"";
+      }
+    }
+
+    std::string lastType{};
+    for (auto const& i : _registry) {
+      std::string name{i.second->name()};
+      std::string alternativeName;
+      if (!v2) {
+        if (v1suppressions.find(name) != v1suppressions.end()) {
+          continue;
+        }
+        // In v1 we do a name conversion. Note that we set
+        // alternativeName == name in the end, in v2 though,
+        // alternativeName is empty and no conversion happens.
+        auto it = nameVersionTable.find(name);
+        if (it != nameVersionTable.end()) {
+          name = it->second;
+        }
+        static std::string const ARANGODB_CONNECTION = "arangodb_connection_";
+        static std::string const POOL_AGENCYCOMM = "pool=\"AgencyComm\"";
+        static std::string const POOL_CLUSTERCOMM = "pool=\"ClusterComm\"";
```

Comment:
I agree with your comments about this file. But are you sure that I should do it in this pr? 
Now I just renamed it, removed the unused code.

From the point of view of refactoring
For example here these constants can be `constexpr std::string_view`.
Or, for example, it seems rather strange to me that each metric builder (DECLARE ..) is a separate class, why not `using` with a compile-time name "string"? Because help "string" is dynamic (
Or, for example, MetricsKey isn't really needed, we can simply write a comparator for the set.
Well, other points that I wrote in descriptions

I don't know if it is worth changing in this PR? 
As I wrote, I tried to make a minimum of changes.

Reply:
It's fine for me to not do the changes in this PR.
But if you are planning larger follow-up refactorings for the metrics after this PR is merged, I would appreciate if you could include the above suggestions there too. Thanks!